### PR TITLE
Feature: Configurable output resolution

### PR DESCRIPTION
**Pull request by tdaugaard**
Add `obsBaseResolution` and `obsOutputResolution` as configurable settings.

Only `obsOutputResolution` can be configured at the moment, and base resolution will always be set to the resolution of the selected monitor.

Both have been made configurable'ish because it is intended to be exposed at a later time with a different mechanic that isn't suitable for this change, and it was easier to add both now. In the future when Game Capture is implemented, we might want to make Base Resolution configurable, so that's why.

Please note, that the changes to output resolution does not always take effect before the application has been restarted. This is a known issue with the reconfiguration of OBS that applies to most of its settings that are saved in `basic.ini`, but _not_ those that are set in-memory like output sources and scenes.

### CODE REVIEW

**Comment by aza547** [comment]
This looks great - not reviewed code yet but had a quick test and mostly worked- one issue that app crashes when doing first time config (I deleted my config-v2.json and tried to reconfigure and it closes when I press the settings). I think we've seen this issue recently, I'll try refresh my memory.

**Comment by aza547** [comment]
> This looks great - not reviewed code yet but had a quick test and mostly worked- one issue that app crashes when doing first time config (I deleted my config-v2.json and tried to reconfigure and it closes when I press the settings). I think we've seen this issue recently, I'll try refresh my memory.

Ah yes, it's the same problem as we had with getting the audio devices before OBS has been configured. I think you want a guard around this before we have a recorder object. For example:

```
ipcMain.on('getAudioDevices', (event) => {
  // We can only get this information if the recorder (OBS) has been
  // initialized and that only happens when the storage directory has
  // been configured.
  if (!recorder) {
    event.returnValue = { input: [], output: [] };
    return;
  }

  event.returnValue = {
    input: getAvailableAudioInputDevices(),
    output: getAvailableAudioOutputDevices(),
  };
});
```

File: src/main/obsRecorder.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { getAvailableAudioInputDevices, getAvailableAudioOutputDevices } from ".
 import { RecorderOptionsType } from "./recorder";
 import { OurDisplayType } from "./types";
 import { Size } from "electron";
+import { IScene } from "obs-studio-node";
```
Comment (aza547) [inline src/main/obsRecorder.ts:7]: I've seen this naming a lot but never understood it. What is the I prefix?

**Comment by tdaugaard** [comment]
Could be, yeah. I have been working on another branch that would allow us to initialize OBS without starting the buffer recording _before_ we have any meaningful configuration that would allow us to gather these values and avoid the guards here, but that is obviously not part of this PR so I'll find a fix for this seperately.

**Review by aza547 — CHANGES_REQUESTED** [review]
No criticisms of the code, just the one bug to fix and one question :)

File: src/main/obsRecorder.ts (hunk 1 above)
Comment (tdaugaard) [reply src/main/obsRecorder.ts:7]: It's a prefix commonly used for interfaces, mostly to be able to visually distinguish them from classes and types.

**Comment by tdaugaard** [comment]
Guard has been added against OBS not being initialized.

File: src/main/logutils.ts [hunk 2]
```diff
@@ -669,6 +669,7 @@ const forceStopRecording = () => {
  * @returns {[boolean, boolean]} retailRunning, classicRunning
  */
 const checkWoWProcess = async (): Promise<[boolean, boolean]> => {
+    return [true, false];
```
Comment (aza547) [inline src/main/logutils.ts:672]: it happened!

File: src/main/main.ts [hunk 3]
```diff
@@ -410,6 +412,16 @@ ipcMain.on('settingsWindow', (event, args) => {
     event.returnValue = getAvailableDisplays();
     return;
   }
+
+  if (args[0] === 'getObsAvailableResolutions') {
+    if (!recorder) {
```
Comment (aza547) [inline src/main/main.ts:417]: pedantic comment, but perhaps neater as:

```
if (args[0] === 'getObsAvailableResolutions') {
  event.returnValue = { 'Base': [], 'Output': [] };
  if (!recorder) return;
  event.returnValue = getObsResolutions();
}
```

File: src/settings/VideoSettings.tsx [hunk 4]
```diff
@@ -22,6 +26,10 @@ export default function VideoSettings() {
     setConfig((prevConfig: any) => ({ ...prevConfig, [stateKey]: value }));
   };
 
+  if (!fpsOptions.includes(config.obsFPS)) {
```
Comment (aza547) [inline src/settings/VideoSettings.tsx:29]: Why is this required? Shouldn't we be getting a default from electron store if it's not set?

File: src/settings/VideoSettings.tsx [hunk 5]
```diff
@@ -46,30 +57,70 @@ export default function VideoSettings() {
...
-            )}
-          </Select>
-        </FormControl>
-        <Tooltip title={configSchema["monitorIndex"].description} >
-          <IconButton>
-            <InfoIcon style={{ color: 'white' }}/>
-          </IconButton>
-        </Tooltip>
-      </Box>
+      <FormControl fullWidth>
```
Comment (aza547) [inline src/settings/VideoSettings.tsx:60]: Seem to have lost the MUI box element here. I think that was crucial for aligning things when I made the settings changes recently. Is that deliberate? Does it still look OK? Also lost some of the margin specification. Suspect a conflict resolution gone the wrong direction?

**Review by aza547 — COMMENTED** [review]
Few more comments.

**Comment by aza547** [comment]
Oh, and need a changelog entry for this please

File: src/main/logutils.ts (hunk 2 above)
Comment (tdaugaard) [reply src/main/logutils.ts:672]: HA HA HA. _Maaaan._ Fixing. 😂

File: src/main/main.ts (hunk 3 above)
Comment (tdaugaard) [reply src/main/main.ts:417]: I'm not a huge fan of non-braces if's, but I see you've done it in _many_ places so as you wish =)

File: src/settings/VideoSettings.tsx (hunk 4 above)
Comment (tdaugaard) [reply src/settings/VideoSettings.tsx:29]: If someone edits their config by hand and types 120 or 144 or something, OBS will just use 60, but in order to show something meaningful in the dropdown. `FPSCommon` only supports a few common FPS values.

File: src/settings/VideoSettings.tsx (hunk 5 above)
Comment (tdaugaard) [reply src/settings/VideoSettings.tsx:60]: That's not inconceivable. I'll have a look at it.

File: src/main/main.ts (hunk 3 above)
Comment (tdaugaard) [reply src/main/main.ts:417]: This was apparently not possible, as setting `event.returnValue` twice resulted in the second assignment being ignored and not returned. I suspect that behind the scens, any assignment to `returnValue` will instantly return the `ipc.sendSync()` call and thus render subsequent assignments pointless.

Comment (aza547) [reply src/main/main.ts:417]: To be honest I'm not sure what I prefer. I spent the first year or two of my career just writing Java as a {} using langauge and the coding standards at work mandate

```
if ()
{
 ...
}
else 
{
  ...
}
```

formatting. Not going to get excited about whatever goes in here.

Comment (aza547) [reply src/main/main.ts:417]: Interesting re the event.returnValue stuff though - I think I actually spotted that once too but obviously forgot. Fine as is!

**Comment by aza547** [comment]
Great! Merging.

File: src/main/main.ts (hunk 3 above)
Comment (tdaugaard) [reply src/main/main.ts:417]: My person preference is that whatever makes the intent of the code easier to read is whatever I go with, but I can see how the following can look off, as the `else` could be easily missed.

```js
if (...) {
} else
if (...) {
}
```
